CLI doesn't run `terraform init` if new module is used

<!--- Please keep this note for the community --->

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### cdktf & Language Versions

`0.15.3`

<!--- Please make note of the version you are using for the `hashicorp/terraform-cdk` and language bindings you are using.
If you are not running the latest version, please upgrade because your issue may have already been fixed. --->

### Affected Resource(s)

Modules

<!--- Please list the affected resources. --->

### Debug Output

```
[2023-02-08T17:22:54.713] [ERROR] default - ╷
│ Error: Module not installed
│
│   on cdk.tf.json line 12, in module:
│   12:     "label": {
│
│ This module is not yet installed. Run "terraform init" to install all
│ modules required by this configuration.
test-cdktf-module-init  ╷
                        │ Error: Module not installed
                        │
                        │   on cdk.tf.json line 12, in module:
                        │   12:     "label": {
                        │
                        │ This module is not yet installed. Run "terraform init" to install all
                        │ modules required by this configuration.
                        ╵

⠏  Processing
Error: External Error: Stack failed to plan: test-cdktf-module-init. Please check the logs for more information.
```

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.
--->

### Expected Behavior

CLI should run `terraform init`

<!--- What should have happened? --->

### Actual Behavior

Does not run `terraform init` which fails

<!--- What actually happened? --->

### Steps to Reproduce

1. Create a new CDKTF project
2. Run `cdktf plan` on the empty project
3. Add a module to `cdktf.json` (e.g. `"cloudposse/label/null@~> 0.25.0"`)
4. Run `cdktf get`
5. Use the module
```ts
import { Label } from './.gen/modules/cloudposse/null/label';
// ...
new Label(this, "label", {});
```
6. Run `cdktf plan`
7. ~Profit~ Error

<!--- Please list the steps required to reproduce the issue. --->

### Important Factoids

Probably introduced by #2572 – We can refactor that fix to still run `init` on every `plan`/`apply` but only run `providers lock` selectively (the latter downloads the re-providers every time which was the real time consuming culprit that we intended to fix in that PR).

<!--- Are there anything atypical about your accounts that we should know? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

--- a/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
+++ b/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
@@ -238,7 +238,12 @@ export class CdktfStack {
     const terraform = await this.terraformClient();
     const needsLockfileUpdate = await this.checkNeedsLockfileUpdate();
     const needsUpgrade = await this.checkNeedsUpgrade();
-    await terraform.init({ needsUpgrade, noColor ?? false, needsLockfileUpdate, this.options.migrateState ?? false });
+    await terraform.init({
+      needsUpgrade,
+      noColor: noColor ?? false,
+      needsLockfileUpdate,
+      migrateState: this.options.migrateState ?? false,
+    });
     return terraform;
   }
 

--- a/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
+++ b/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
@@ -236,18 +236,9 @@ export class CdktfStack {
 
   public async initalizeTerraform(noColor?: boolean) {
     const terraform = await this.terraformClient();
-    const needsInit = await this.checkNeedsInit();
-    if (!needsInit) {
-      // Skip terraform init as everything's up to date
-      return terraform;
-    }
-
+    const needsLockfileUpdate = await this.checkNeedsLockfileUpdate();
     const needsUpgrade = await this.checkNeedsUpgrade();
-    await terraform.init(
-      needsUpgrade,
-      noColor ?? false,
-      this.options.migrateState ?? false
-    );
+    await terraform.init({ needsUpgrade, noColor ?? false, needsLockfileUpdate, this.options.migrateState ?? false });
     return terraform;
   }
 
@@ -262,12 +253,11 @@ export class CdktfStack {
     }, {} as Record<string, ProviderConstraint>);
   }
 
-  private async checkNeedsInit(): Promise<boolean> {
+  private async checkNeedsLockfileUpdate(): Promise<boolean> {
     if (this.options.migrateState) {
       // If we're migrating state, we need to init
       return true;
     }
-
     const lock = new TerraformProviderLock(this.stack.workingDirectory);
     const lockFileExists = await lock.hasProviderLockFile();
 

--- a/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
@@ -106,18 +106,19 @@ export class TerraformCli implements Terraform {
         createTerraformLogHandler(phase, filter)(stderr.toString(), true);
   }
 
-  public async init(
-    needsUpgrade: boolean,
-    noColor: boolean,
-    migrateState: boolean
-  ): Promise<void> {
+  public async init(opts: {
+    needsUpgrade: boolean;
+    noColor?: boolean;
+    migrateState: boolean;
+    needsLockfileUpdate: boolean;
+  }): Promise<void> {
     await this.setUserAgent();
 
     const args = ["init"];
-    if (needsUpgrade) {
+    if (opts.needsUpgrade) {
       args.push("-upgrade");
     }
-    if (noColor) {
+    if (opts.noColor) {
       args.push("-no-color");
     }
 
@@ -140,7 +141,7 @@ export class TerraformCli implements Terraform {
       (data) => {
         stdout(data);
         if (data.includes("Should Terraform migrate your existing state?")) {
-          if (migrateState) {
+          if (opts.migrateState) {
             actions.writeLine("yes");
           } else {
             actions.stop();
@@ -165,22 +166,24 @@ export class TerraformCli implements Terraform {
     // TODO: this might have performance implications because we don't know if we're
     // running a remote plan or a local one (so we run it always for all platforms)
     // while we'd only need it for remote plans
-    await exec(
-      terraformBinaryName,
-      [
-        "providers",
-        "lock",
-        "-platform=linux_amd64",
-        ...(noColor ? ["-no-color"] : []),
-      ],
-      {
-        cwd: this.workdir,
-        env: process.env,
-        signal: this.abortSignal,
-      },
-      this.onStdout("init"),
-      this.onStderr("init")
-    );
+    if (opts.needsLockfileUpdate) {
+      await exec(
+        terraformBinaryName,
+        [
+          "providers",
+          "lock",
+          "-platform=linux_amd64",
+          ...(opts.noColor ? ["-no-color"] : []),
+        ],
+        {
+          cwd: this.workdir,
+          env: process.env,
+          signal: this.abortSignal,
+        },
+        this.onStdout("init"),
+        this.onStderr("init")
+      );
+    }
   }
 
   private get isCloudStack(): boolean {

--- a/packages/@cdktf/cli-core/src/lib/models/terraform.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform.ts
@@ -129,11 +129,13 @@ export type TerraformDeployState =
   | { type: "external sentinel override reply"; overridden: boolean };
 
 export interface Terraform {
-  init: (
-    upgrade: boolean,
-    noColor: boolean,
-    migrateState: boolean
-  ) => Promise<void>;
+  init: (opts: {
+    needsUpgrade: boolean;
+    noColor?: boolean;
+    needsLockfileUpdate: boolean;
+    migrateState: boolean;
+  }) => Promise<void>;
+
   plan: (opts: {
     destroy: boolean;
     refreshOnly?: boolean;

--- a/test/typescript/init-repeatedly/cdktf.json
+++ b/test/typescript/init-repeatedly/cdktf.json
@@ -1,8 +1,12 @@
 {
   "language": "typescript",
   "app": "npm run --silent compile && node main.js",
-  "terraformProviders": [
-    "null"
+  "terraformProviders": ["null"],
+  "terraformModules": [
+    {
+      "name": "our-local-module",
+      "source": "./local-module"
+    }
   ],
   "context": {
     "excludeStackIdFromLogicalIds": "true"

--- a/test/typescript/init-repeatedly/cdktf.json
+++ b/test/typescript/init-repeatedly/cdktf.json
@@ -1,0 +1,10 @@
+{
+  "language": "typescript",
+  "app": "npm run --silent compile && node main.js",
+  "terraformProviders": [
+    "null"
+  ],
+  "context": {
+    "excludeStackIdFromLogicalIds": "true"
+  }
+}

--- a/test/typescript/init-repeatedly/local-module/main.tf
+++ b/test/typescript/init-repeatedly/local-module/main.tf
@@ -1,0 +1,18 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: MPL-2.0
+
+variable "enabled" {
+  description = "Flag to optionally disable usage of this module."
+  type        = bool
+  default     = true
+}
+
+variable "set" {
+  description = "Making sure bindings can be created for modules using the set type"
+  type = set(string)
+}
+
+output "create_cmd_bin" {
+  description = "The full bin path & command used on create"
+  value       = "foo"
+}

--- a/test/typescript/init-repeatedly/main.ts
+++ b/test/typescript/init-repeatedly/main.ts
@@ -1,0 +1,22 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import { Construct } from "constructs";
+import { App, TerraformStack, Testing, LocalBackend } from "cdktf";
+import { OurLocalModule } from "./.gen/modules/our-local-module";
+
+export class StackWithModule extends TerraformStack {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+    new LocalBackend(this, {
+      path: "terraform.tfstate",
+    });
+
+    new OurLocalModule(this, "local-module", {
+      set: ["test", "sets"],
+    });
+  }
+}
+
+const app = Testing.stubVersion(new App({ stackTraces: false }));
+new StackWithModule(app, "repeated-init-stack");
+app.synth();

--- a/test/typescript/init-repeatedly/test.ts
+++ b/test/typescript/init-repeatedly/test.ts
@@ -1,8 +1,8 @@
 // Copyright (c) HashiCorp, Inc
 // SPDX-License-Identifier: MPL-2.0
 import { TestDriver, onPosix } from "../../test-helper";
-import path from "node:path";
-import fs from "fs-extra";
+import * as path from "path";
+import * as fs from "fs-extra";
 
 // Since chalk auto-detects the capabilities, colored output is
 // deactivated by default in non-tty environments. We had an issue
@@ -24,8 +24,13 @@ describe("test with colors", () => {
     await driver.get();
   });
 
-  it("doesn't repeat init on subsequent diff", async () => {
-    const output = await driver.synth();
-    console.log(output);
+  it("repeats init on subsequent diff, but not provider lock", async () => {
+    const output = await driver.diff();
+    expect(output).toContain("Initializing provider plugins");
+    expect(output).toContain("validated the lock file");
+
+    const secondDiffOutput = await driver.diff();
+    expect(secondDiffOutput).toContain("Initializing provider plugins");
+    expect(secondDiffOutput).not.toContain("validated the lock file");
   });
 });

--- a/test/typescript/init-repeatedly/test.ts
+++ b/test/typescript/init-repeatedly/test.ts
@@ -1,0 +1,31 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import { TestDriver, onPosix } from "../../test-helper";
+import path from "node:path";
+import fs from "fs-extra";
+
+// Since chalk auto-detects the capabilities, colored output is
+// deactivated by default in non-tty environments. We had an issue
+// in the past, where this turned out to be a problem
+// see https://github.com/hashicorp/terraform-cdk/issues/139
+
+describe("test with colors", () => {
+  let driver: TestDriver;
+
+  beforeAll(async () => {
+    driver = new TestDriver(__dirname);
+    driver.switchToTempDir();
+    await driver.init("typescript");
+    driver.copyFiles("main.ts", "cdktf.json");
+    fs.copySync(
+      path.join(__dirname, "local-module"),
+      path.join(driver.workingDirectory, "local-module")
+    );
+    await driver.get();
+  });
+
+  it("doesn't repeat init on subsequent diff", async () => {
+    const output = await driver.synth();
+    console.log(output);
+  });
+});

